recipes/lora_finetune_single_device.py

```diff
@@ -415,6 +429,22 @@ def _setup_model(
             self.adapter_params.items(), dtype=self._dtype
         )
 
+        self.activations_handling_ctx = contextlib.nullcontext()
+        if enable_activation_offloading:
+            self.activations_handling_ctx = OffloadActivations(
+                use_streams=offload_with_streams
+            )
+
+            # Below is our hack to disable offloading the last output Linear in every
+            # step, as the cost for offloading the activation and then soon after bringing
+            # it back is expensive. Moreover, due to heuristics in our streaming API,
+            # we actually use more memory if we offload it as it interferes with chunkedCE.
+            noop_ctx = NoOpManager()
+            model.output.register_forward_pre_hook(lambda *args: noop_ctx.__enter__())
+            model.output.register_forward_hook(
+                lambda *args: noop_ctx.__exit__(), always_call=True
```

Comment:
Ah yea, I agree we need to test these. The hooks do only work with module (since they're module hooks) so that would require some design...e.g., we enable adding hooks to this class you added, or we make it an nn.Module still but exclude it from FSDP/other wrappers. Or something else 😛

Reply:
Oh yeah this tied embedding business is actually a pretty important point. As a hack we can maybe do something like `if hasattr(model, "output") and isinstance(model.output, nn.Module)` to gate the registration of these hooks? 

Would have to think more about whether we can get away with making `TiedLinear` an nn.Module though.. half the reason we're doing things this way is to avoid the breaking of references for tied weights when FSDP calls `to_empty`. Is there an easy way to exclude an nn.Module from all FSDP's usual hook registration etc?

Anyways before I get too far down that rabbit hole, I'd propose just going with the hack (provided it works). It's dumb, simple, and explicit, which I like.